Simplify dockerCmd

vdemeester opened the pull request:
Looking back at #14603, there is still something that felt weird : we are using `dockerCmd` 99% of the time like this :

``` go
out, _ := dockerCmd(c, "ps")
```

The reason is the following : we don't really care about the error code in that case, because how it's written it will always be 0 if the method _passes_. If we look at https://github.com/docker/docker/blob/master/integration-cli/utils.go#L60 the error code will not be 0 if there is an error — and if there is a error, the assert is _KO_ and the test will fail. 🐧

I might miss something, and in that case feel free to close this one :wink:. 
🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

tiborvass commented:
I don't even know why or when was `dockerCmd` introduced, but I would like to revisit all those functions including `runCommand*`.

Facts:
- we almost **never** need an exit status, `out` and `err` are enough.
- we almost **never** make the distinction between stdout and stderr, so `out` is sufficient for output.

`out, err := runCmd(...)` and `out := dockerCmd(...)`. All the other cases should use `exec.Command`

What do you guys think? @vdemeester do you agree? Ping @LK4D4 @unclejack

tiborvass commented:
Either way, I like this PR so +1 to this, if nobody agrees with what i said above.

unclejack commented:
The utilities from integration-cli were meant to be sort of a standard library for the integration-cli tests which helps avoid extreme duplication of the code. Some of those utility functions were rewritten a few times before the first version of integration-cli was sent as a PR and checked into the code.

Changing the code to make it work the way it needs to work today makes sense and this has already been done a few times. However, we need to make sure we do so safely and that we don't remove too much of the integration-cli library of utility functions in order to avoid code duplication for future tests.

We should really test these utility functions to make sure that they do what they were meant to do. Trivial changes might turn out to break things and we've already got problems with integration-cli tests failing sometimes.

Instead of merging this right away, I'd strongly suggest that we add tests for the integration-cli utility functions first. To be more specific, here are the kinds of tests we should have:
- tests to ensure error handling works as expected
- tests to check that the functions do what they're meant to do
- test the functions with the testing framework to ensure that they behave the way we assume they behave in the context of integration-cli

Overall, I agree with doing this, just not before we add proper tests for to the integration-cli utility functions themselves.

duglin commented:
tests of the test tools - I love it! :-)  +1

vdemeester commented:
There is no rush on that, so yeah I think we can do better than just the PR as is :blush:.

@tiborvass I agree for the most part. There is a few tests where we actually test out the ``exitCode``, like [here](https://github.com/docker/docker/blob/master/integration-cli/docker_cli_run_unix_test.go#L292) so we might want to be able to get it somehow (or process the error return to get the exitCode). That's the same with ``stdout``/``stderr``, there is a few *edge case* when we want them separate.

@unclejack @duglin I like the idea of testing the *test utility* function — I'll take a look at that :wink:.

duglin commented:
It would be nice if dockerCmd used dockerCmdWithError (assuming that's the right func) but then just ignored the exit code. Then we have one central func to invoke docker but do different things with the results.  Avoids code dup and fragmentation.

tiborvass commented:
@vdemeester @duglin i find it simpler that for edgecases we just use the plain old exec.Command that works perfectly well but maybe it's just me.

abronan commented:
Sorry @vdemeester, needs a rebase

vdemeester commented:
@abronan I know :blush:. I'm hesitant to do it right now because, with the discussion we had, this change is not the first we should do. There is consolidation to do one the integration-cli tests: 
- consolidate the integration-cli tests utils (tests),
- then refactoring to ease the use even more when we are confident we don't break

I'm wondering out to proceed too : leave this open (but not mergeable) to have something that says "there is some work on this", or close it, write an issue to list the work to be done and close this one. 

In the end it's pretty sure there is other PRs that will be needed before this one should be merge so it will end up being closed and carried or rewrote :blush:. WDYT ? :sweat_smile:

cpuguy83 commented:
I think it would suffice to go ahead and cleanup the usage of `dockerCmd` since the status code is not needed at all, and `dockerCmdWithError`.
I don't think we really need `dockerCmdWithError` since it is less flexible than exec.Command and provides no real utility over it.

vdemeester commented:
I've rebased it against master so it's *mergeable*.

To consolidate a bit the `integration-cli` tests, here is what I propose (and willing to work on :wink:) :
- Move *utils* (`utils.go` and `docker_utils.go`) to a package. That way we can **test it** and **export only what we want to be used** in the `integration-cli`. This would mean using `DockerCmd`, `BuildImage`, etc... The unit test of the `integration-cli/utils` package would be run just before `test-integration-cli`.
- Add tests to all exported function (`RunCommandWithOutput`, etc..).
- Simplify the number of function used and their signatures : We either want the command to be successful and get the output, or the command to not be successful and get everything from it (output, error, exitcode, …).
- Add some `Checker` to make even more used of the `check` library. And *over time*, using more `c.Assert` for the most common case (like `!=`, …).

This work would be separate in several pull-requests in order. So for this PR we could either merged it as is (makeing `dockerCmd` simple) or close it and these changes would take place in an other PR (or do all the thing in this PR but it feels like a lot of changes :sweat_smile:).

@cpuguy83 @tiborvass @duglin @unclejack WDYT ?

unclejack commented:
@vdemeester Moving the integration-cli utils to a package is a bad idea because the integration-cli code has to be changed throughout the code base and that breaks a lot of PRs.

Is there a problem with having the unit tests for integration-cli in the same places and running those first?
Either way, we should add tests first before refactoring anything. Simple and seemingly innocuous changes have caused troubles in the past. integration-cli is the last place where I'd start making changes throughout the code base without having tests first.

vdemeester commented:
> Is there a problem with having the unit tests for integration-cli in the same places and running those first?

I'm investigating to make it as smooth as possible but, to run unit-tests on `integration-cli` :

```bash
$ DOCKER_BINARY="../bundles/latest/binary/docker" go test docker_test_vars.go utils_test.go utils.go
INFO: Testing against a local daemon
--- FAIL: TestRunCommandWithOutput (0.00s)
        utils_test.go:15: hola
FAIL
FAIL    command-line-arguments  0.008s

$ DOCKER_BINARY="../bundles/latest/binary/docker" go test docker_test_vars.go utils_test.go utils.go
ok      command-line-arguments  0.012s
```

If we don't specify the `*.go` files, this will try to pick up the integration one. `docker_test_vars.go` is there for the `dockerBinary` variable. It's not that bad either but I though it would be easier to set-up a package just for theses.

> Moving the integration-cli utils to a package is a bad idea because the integration-cli code has to be changed throughout the code base and that breaks a lot of PRs.

Yeah, that's true. I'll go with the above solution for now and we'll see later (at least we will have unit-tests). I'll work on the unit-tests first, probably making a PR just for that, and then we'll be back on this one. @unclejack sounds good ? :wink:

vdemeester commented:
@duglin @tiborvass @unclejack I started to work on this here : https://github.com/docker/docker/compare/master...vdemeester:integration-cli-utils-tests. 

It's gonna be a little bit more tricky to test `docker_utils.go` as it depends on the `docker command` (and a docker engine to run). I'll try to mock the `docker` binary somehow to test these.

unclejack commented:
@vdemeester Sounds good, the `docker` binary mocking as well.

tiborvass commented:
@vdemeester sorry needs a rebase :S

vdemeester committed:
Simplify dockerCmd

dockerCmd now only returns the output, as the status can't be
anything else than 0 — if there is an error, the assert will be
false and the test will stop anyway.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
Lol a conflict on almost every files :sweat_smile:.. some heavy changes going on in `integration-cli` :smile_cat:.

icecrime commented:
@vdemeester So should we close this one? :-(

vdemeester commented:
@icecrime you're right, I'm gonna close it for now, and I'll re-open a new one when `dockerCmd` is covered :wink:
